source/core_mqtt_serializer.c

```diff
@@ -1785,896 +3414,1985 @@ MQTTStatus_t MQTT_GetConnectPacketSize( const MQTTConnectInfo_t * pConnectInfo,
+            pIndex[1] = UINT32_BYTE2(sessionExpiry);
+            pIndex[2] = UINT32_BYTE1(sessionExpiry);
+            pIndex[3] = UINT32_BYTE0(sessionExpiry);
+            pIndex = &pIndex[4];
+            pPropertyBuilder->fieldSet = UINT32_SET_BIT(pPropertyBuilder->fieldSet, MQTT_SESSION_EXPIRY_INTERVAL_POS);
+            pPropertyBuilder->currentIndex += 5;
+        }
+        return status;
     }
-    else if( pFixedBuffer->pBuffer == NULL )
+
+    MQTTStatus_t MQTTPropAdd_ConnReceiveMax(MqttPropBuilder_t* pPropertyBuilder, uint16_t receiveMax)
     {
-        LogError( ( "pFixedBuffer->pBuffer cannot be NULL." ) );
-        status = MQTTBadParameter;
+        MQTTStatus_t status = MQTTSuccess;
+        if ((pPropertyBuilder == NULL))
+        {
+            LogError(("Arguments cannot be NULL : pPropertyBuilder=%p.", (void*)pPropertyBuilder));
+            status = MQTTBadParameter;
+        }
+        else if (pPropertyBuilder->pBuffer == NULL)
+        {
+            LogError(("Arguments cannot be NULL : pPropertyBuilder->pBuffer=%p.", (void*)pPropertyBuilder->pBuffer));
+            status = MQTTBadParameter;
+        }
+        else if ((receiveMax == 0) || (UINT32_CHECK_BIT(pPropertyBuilder->fieldSet, MQTT_RECEIVE_MAXIMUM_POS)))
+        {
+            LogError(("Invalid arguments passed to MQTTPropAdd_ConnReceiveMax."));
+            status = MQTTBadParameter;
+        }
+        else
+        {
+            uint8_t* pIndex = pPropertyBuilder->pBuffer + pPropertyBuilder->currentIndex;
+            *pIndex = MQTT_RECEIVE_MAX_ID;
+            pIndex++;
+            pIndex[0] = UINT16_HIGH_BYTE(receiveMax);
+            pIndex[1] = UINT16_LOW_BYTE(receiveMax);
+            pIndex = &pIndex[2];
+            pPropertyBuilder->fieldSet = UINT32_SET_BIT(pPropertyBuilder->fieldSet, MQTT_RECEIVE_MAXIMUM_POS);
+            pPropertyBuilder->currentIndex += 3;
+        }
+        return status;
     }
-    else
+
+    MQTTStatus_t MQTTPropAdd_ConnMaxPacketSize(MqttPropBuilder_t* pPropertyBuilder, uint32_t maxPacketSize)
     {
-        /* Empty else MISRA 15.7 */
+        MQTTStatus_t status = MQTTSuccess;
+        uint8_t* pIndex;
+        if ((pPropertyBuilder == NULL))
+        {
+            LogError(("Arguments cannot be NULL : pPropertyBuilder=%p.", (void*)pPropertyBuilder));
+            status = MQTTBadParameter;
+        }
+        else if (pPropertyBuilder->pBuffer == NULL)
+        {
+            LogError(("Arguments cannot be NULL : pPropertyBuilder->pBuffer=%p.", (void*)pPropertyBuilder->pBuffer));
+            status = MQTTBadParameter;
+        }
+        else if ((maxPacketSize == 0) || (UINT32_CHECK_BIT(pPropertyBuilder->fieldSet, MQTT_MAX_PACKET_SIZE_POS)))
+        {
+            LogError(("Max packet size already set"));
+            status = MQTTBadParameter;
+        }
+        else
+        {
+            pIndex = pPropertyBuilder->pBuffer + pPropertyBuilder->currentIndex;
+            *pIndex = MQTT_MAX_PACKET_SIZE_ID;
```

Comment:
We need checks to make sure that we are not going out of bound of the `pPropertyBuilder->pBuffer` by verifying `pPropertyBuilder->currentIndex` and the required encoding length with `pPropertyBuilder->bufferLength`.
This can be added as a separate preliminary check for all similar APIs with a separate error code.